### PR TITLE
pageserver: fix eviction across generations

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

Bug was introduced by me in 83ae2bd82c

When eviction constructs a RemoteLayer to replace the layer it just evicted, it is building a LayerFileMetadata using its _current_ generation, rather than the generation of the layer.

## Summary of changes

- Retrieve Generation from RemoteTimelineClient when evicting.  This will no longer be necessary when #4938 lands.
- Add a test for the scenario in question (this fails without the fix).

### CODE REVIEW

File: pageserver/src/tenant/remote_timeline_client.rs [hunk 1]
```diff
@@ -501,14 +501,20 @@ impl RemoteTimelineClient {
...
-        layer_metadata: &LayerFileMetadata,
     ) -> anyhow::Result<u64> {
+        let layer_metadata = {
+            let q = self.upload_queue.lock().unwrap();
+            q.get_layer_metadata(layer_file_name)?.ok_or_else(|| {
+                anyhow::anyhow!(
+                    "Tried to download layer file that does not exist in remote metadata"
+                )
+            })?
+        };
```
Comment (problame) [inline pageserver/src/tenant/remote_timeline_client.rs:516]: Technically, we need a `wait_completion()` before calling `get_layer_metadata`.
Reason: we update `latest_files` early, when we schedule the upload, not when the upload is completed.
I guess it doesn't matter in practice because we should never be asking to download a layer that hasn't been uploaded yet.

**Review by problame — COMMENTED** [review]
I think removing the `layer_metadata` from `RemoteLayer` is moving things in the right direction.

Effectively, the layer map along with the layers objects it references is the runtime representation of the `index_part.json`.

What this PR does is to move the "which generation number does a layer have" piece of information out of layer map / its layer objects into the upload_queue.rs.

I think it would be cleaner to _add_ `layer_metadata` fields to all the other layer types.
As an immutable field.
For layer uploads/ download / eviction, we'd then simply pass that field to remote_timeline_client.

File: pageserver/src/tenant/remote_timeline_client.rs (hunk 1 above)
Comment (jcsp) [reply pageserver/src/tenant/remote_timeline_client.rs:516]: > I guess it doesn't matter in practice because we should never be asking to download a layer that hasn't been uploaded yet.

Right, that's the intent.

**Comment by jcsp** [comment]
>I think it would be cleaner to add layer_metadata fields to all the other layer types.
> As an immutable field.
> For layer uploads/ download / eviction, we'd then simply pass that field to remote_timeline_client.

I think we can get there once we have the invariant that local layers are only uploaded in the same process lifetime that they were created (then we can stamp an immutable LayerMetadata that has the correct generation).

However, we do also have the size already in `PersistentLayerDesc`, so it might make sense to think of that one as the immutable part for local layer state, and aim toward an end state where our local and remote metadata are explicitly separate, rather than having local layer objects carry around a generation at all.

**Comment by problame** [comment]
> I think we can get there once we have the invariant that local layers are only uploaded in the same process lifetime that they were created (then we can stamp an immutable LayerMetadata that has the correct generation).

That's ready for merge, I think: https://github.com/neondatabase/neon/pull/5198

**Comment by koivunej** [comment]
I may have moved it there, or am just keeping it with the layer. With my big layer PR it will be the load time value anyways. Please do not make extra changes.

**Comment by jcsp** [comment]
Looking at the future state in #4938 , the generation is carried around inside `LayerInner`, so this issue of it getting re-created during eviction doesn't occur.

I'm going to leave this PR open because the test is still useful, and so that we don't forget.

**Comment by jcsp** [comment]
Updated to avoid messing with layer types -- the fix is kind of verbose for what it is, but all the new get_layer_metadata stuff on RemoteTimelineClient can be reverted  once #4938 merges.  In the meantime this fixes the bug and unblocks more testing on generations/locationconf stuff.

File: pageserver/src/tenant/timeline.rs [hunk 2]
```diff
@@ -1294,7 +1294,23 @@ impl Timeline {
...
+            .map_err(EvictionError::LayerNotFound)?
+            .ok_or_else(|| {
+                EvictionError::LayerNotFound(anyhow::anyhow!("Layer not in remote metadata"))
+            })?;
+        if layer_metadata.file_size() != layer_file_size {
+            return Err(EvictionError::MetadataInconsistency(format!(
+                "Layer size {layer_file_size} doesn't match remote metadata file size {}",
+                layer_metadata.file_size()
+            )));
+        }
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:1313]: Why did you add this check?
We're deleting the `local_layer` file anyway, why bother checking?
We didn't check before.

File: test_runner/regress/test_pageserver_generations.py [hunk 3]
```diff
@@ -467,3 +483,48 @@ def test_emergency_mode(neon_env_builder: NeonEnvBuilder, pg_bin: PgBin):
...
     assert get_deletion_queue_executed(ps_http) > 0
+
+
+def evict_all_layers(env: NeonEnv, tenant_id: TenantId, timeline_id: TimelineId):
+    timeline_path = env.pageserver.timeline_dir(tenant_id, timeline_id)
+    initial_local_layers = sorted(
+        list(filter(lambda path: path.name != "metadata", timeline_path.glob("*")))
+    )
+    client = env.pageserver.http_client()
+    for layer in initial_local_layers:
```
Comment (problame) [inline test_runner/regress/test_pageserver_generations.py:494]: Instead of scanning the filesystem, you can use the HTTP client to get the layer map dump, then use the `remote` field to figure out whether a layer is local or remote.

This function could then also just live in the http client.

File: pageserver/src/tenant/timeline.rs (hunk 2 above)
Comment (jcsp) [reply pageserver/src/tenant/timeline.rs:1313]: Any time I have two numbers in hand that are _meant_ to be the same, it's a nice opportunity to make sure the assumption is correct.

**Review by problame — APPROVED** [review]
